mlir_graphblas/algorithms.py

```diff
@@ -369,3 +368,143 @@ def vertex_nomination(
 
     node_of_interest = _vertex_nomination(graph, nodes_of_interest)
     return node_of_interest
+
+
+_pagerank = None
+
+
+def pagerank(
+    graph: MLIRSparseTensor, damping=0.85, tol=1e-6, *, maxiter=100
+) -> MLIRSparseTensor:
+    global _pagerank
+    if _pagerank is None:
+        csr64 = SparseEncodingType(["dense", "compressed"], [0, 1], 64, 64)
+        csc64 = SparseEncodingType(["dense", "compressed"], [1, 0], 64, 64)
+        csx64 = SparseEncodingType(["dense", "compressed"], None, 64, 64)
+        cv64 = SparseEncodingType(["compressed"], None, 64, 64)
+        aliases = AliasMap()
+        aliases["CSR64"] = csr64
+        aliases["CSC64"] = csc64
+        aliases["CSX64"] = csx64
+        aliases["CV64"] = cv64
+
+        irb = MLIRFunctionBuilder(
+            "pagerank",
+            input_types=["tensor<?x?xf64, #CSR64>", "f64", "f64", "index"],
+            return_types=["tensor<?xf64, #CV64>", "index"],
+            aliases=aliases,
+        )
+        (A, var_damping, var_tol, var_maxiter) = irb.inputs
+
+        nrows = irb.graphblas.num_rows(A)
+        nrows_i64 = irb.index_cast(nrows, "i64")
+        nrows_f64 = irb.sitofp(nrows_i64, "f64")
+
+        c0 = irb.constant(0, "index")
+        c1 = irb.constant(1, "index")
+        cf1 = irb.constant(1.0, "f64")
+        teleport = irb.subf(cf1, var_damping)
+        teleport = irb.divf(teleport, nrows_f64)
+
+        row_degree = irb.graphblas.reduce_to_vector(A, "count", axis=1)
+        # prescale row_degree with damping factor, so it isn't done each iteration
+        row_degree = irb.graphblas.apply(row_degree, "div", right=var_damping)
+
+        # Use row_degree as a convenient vector to duplicate for starting score
+        r = irb.graphblas.dup(row_degree)
+
+        # r = 1/nrows
+        nrows_inv = irb.divf(cf1, nrows_f64)
+        starting = irb.graphblas.apply(r, "second", right=nrows_inv)
+        starting_ptr8 = irb.util.tensor_to_ptr8(starting)
+
+        # Pagerank iterations
+        rdiff = irb.new_var("f64")
+        prev_score_ptr8 = irb.new_var("!llvm.ptr<i8>")
+        iter_count = irb.new_var("index")
+        with irb.for_loop(
+            0,
+            var_maxiter,
+            iter_vars=[
+                (rdiff, cf1),
+                (prev_score_ptr8, starting_ptr8),
+                (iter_count, c0),
+            ],
+        ) as for_vars:
+            converged = irb.new_var("i1")
+            irb.add_statement(
```

Comment:
There are a few calls to `irb.add_statement` here. We don't have to handle them now, but we should at least open issues to add IR builder ops for `cmpf` and `scf.if` so that we don't forget.

Reply:
I'll go ahead and add an Op for `CmpF` and an issue for both `scf.if` and `scf.while` (used in SSSP)

Reply:
https://github.com/metagraph-dev/mlir-graphblas/issues/159